### PR TITLE
feat: Add DialCache for key-based connection caching

### PR DESCRIPTION
**Pull request by kylecarbs**

### CODE REVIEW

File: wsnet/cache.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+type DialerCache struct {
+	dialerFunc  dialerFunc
+	ttl         time.Duration
+	flightGroup *singleflight.Group
+
+	closed  chan struct{}
+	mut     sync.RWMutex
+	dialers map[string]*Dialer
+	atime   map[string]time.Time
```
Comment (deansheather) [inline wsnet/cache.go:38]: You should use a struct instead of having two maps for the same key.

File: wsnet/cache.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+		dialer := dialer
+		go func() {
+			defer wg.Done()
+
+			evict := false
+			select {
+			case <-dialer.Closed():
+				evict = true
+			default:
+			}
```
Comment (deansheather) [inline wsnet/cache.go:71]:
```suggestion
```

File: wsnet/cache.go [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+// Close closes all cached dialers.
+func (d *DialerCache) Close() error {
+	d.mut.Lock()
+	defer d.mut.Unlock()
+
+	for key, dialer := range d.dialers {
+		d.flightGroup.Forget(key)
+
```
Comment (deansheather) [inline wsnet/cache.go:144]:
```suggestion
```

File: wsnet/cache.go [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+		err := dialer.Close()
+		if err != nil {
+			return err
+		}
+	}
+	close(d.closed)
```
Comment (deansheather) [inline wsnet/cache.go:150]: You should check if it's closed at the top of Dial() and after the flightGroup.Do

File: wsnet/cache_test.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+		require.NoError(t, err)
+		defer l.Close()
+
+		cache := DialCache(time.Hour, func(ctx context.Context, key string) (*Dialer, error) {
+			return DialWebsocket(ctx, connectAddr, nil)
+		})
+		_, cached, err := cache.Dial(context.Background(), "example")
+		require.NoError(t, err)
+		require.Equal(t, cached, false)
+		_, cached, err = cache.Dial(context.Background(), "example")
```
Comment (deansheather) [inline wsnet/cache_test.go]: Check that the returned dialers are `assert.Same()`

File: wsnet/cache_test.go [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+		cache := DialCache(time.Hour, func(ctx context.Context, key string) (*Dialer, error) {
+			return DialWebsocket(ctx, connectAddr, nil)
+		})
+
+		conn, cached, err := cache.Dial(context.Background(), "example")
+		require.NoError(t, err)
+		require.Equal(t, cached, false)
+		require.NoError(t, conn.Close())
+		_, cached, err = cache.Dial(context.Background(), "example")
```
Comment (deansheather) [inline wsnet/cache_test.go]: Check that they aren't the `assert.Same()`

File: wsnet/cache.go [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+			if !evict {
+				ctx, cancel := context.WithTimeout(context.Background(), time.Second*15)
+				defer cancel()
+				err := dialer.Ping(ctx)
+				if err != nil {
+					evict = true
+				}
+			}
+
+			if evict {
```
[the remark this answers is not in the document]
Comment (kylecarbs) [reply wsnet/cache.go:84]: Huh?

File: wsnet/cache.go [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+				_ = dialer.Close()
+				d.mut.Lock()
+				delete(d.atime, key)
+				delete(d.dialers, key)
+				d.mut.Unlock()
```
[the remark this answers is not in the document]
Comment (kylecarbs) [reply wsnet/cache.go:89]: I don't _think_ so. This is inside a WaitGroup goroutine.

File: wsnet/cache.go [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+				if err != nil {
+					evict = true
+				}
+			}
+
+			if evict {
+				_ = dialer.Close()
+				d.mut.Lock()
+				delete(d.atime, key)
+				delete(d.dialers, key)
```
[the remark this answers is not in the document]
Comment (kylecarbs) [reply wsnet/cache.go:88]: Ahh interesting. How should I do that?

File: wsnet/cache.go (hunk 2 above)
Comment (kylecarbs) [reply wsnet/cache.go:71]: Why remove this? We want to evict closed connections.

File: wsnet/cache.go (hunk 1 above)
Comment (kylecarbs) [reply wsnet/cache.go:36]: I felt like it was a bit overkill here... I'll add a comment denoting they're the same.

File: wsnet/cache.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (deansheather) [reply wsnet/cache.go:84]: Use `else` instead of two opposite `if`s

File: wsnet/cache.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (deansheather) [reply wsnet/cache.go:88]: Check if `d.dialers[key]` exists, and is equal to the current dialer pointer you have, if it isn't then don't delete anything

File: wsnet/cache.go (hunk 2 above)
Comment (deansheather) [reply wsnet/cache.go:70]: oops I thought this was d.Closed my bad

File: wsnet/cache.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (deansheather) [reply wsnet/cache.go:89]: This is the last statement in the goroutine so it's effectively the same. You get protection from mut getting left locked from panics if you use defer, so it's better to use defer wherever possible

File: wsnet/cache.go [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+	return dc
+}
+
+type DialerCache struct {
+	ttl         time.Duration
+	flightGroup *singleflight.Group
+	closed      chan struct{}
+	mut         sync.RWMutex
+
+	// Key is the "key" of a dialer.
```
Comment (deansheather) [inline wsnet/cache.go:35]:
```suggestion
	// Key is the "key" of a dialer, which is usually the workspace ID.
```

File: wsnet/cache_test.go [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+		cache := DialCache(0)
+
+		_, cached, err := cache.Dial(context.Background(), "example", dialFunc(connectAddr))
+		require.NoError(t, err)
+		require.Equal(t, cached, false)
+		cache.evict()
+		_, cached, err = cache.Dial(context.Background(), "example", dialFunc(connectAddr))
+		require.NoError(t, err)
+		require.Equal(t, cached, false)
```
Comment (deansheather) [inline wsnet/cache_test.go:68]: Check NotSame here too

**Review by deansheather — APPROVED** [review]

File: wsnet/cache.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (kylecarbs) [reply wsnet/cache.go:84]: Ahh the other one is to ping only if can't evict.

**Pull request merged**